src/edit/index.vue

```diff
@@ -38,73 +45,77 @@ export default {
     isFocused: { type: Boolean, default: false },
     isDisabled: { type: Boolean, default: false }
   },
-  data: () => ({ file: null }),
+  data: () => ({ file: null, error: null }),
   computed: {
     isEmpty() {
-      const { error, fileName } = this.element.data;
+      const { error, fileName } = this.element.data.video;
       return !error && !fileName;
     },
     didUploadFail() {
-      const { status } = this.element.data;
+      const { status } = this.element.data.video;
       return status === ELEMENT_STATE.UPLOADING && !this.file;
     },
     errorMessage() {
-      const { error } = this.element.data;
+      const { error } = this.element.data.video;
       return this.didUploadFail ? UPLOAD_FAILED_ERROR_MSG : error;
     },
     infoMessage() {
-      const { status, playable } = this.element.data;
+      const { status, playable } = this.element.data.video;
       if (status === ELEMENT_STATE.UPLOADING) return UPLOADING_MSG;
       return !playable && PROCESSING_MSG;
     },
     isReadyToUpload() {
-      const { token, uploadUrl } = this.element.data;
+      const { token, uploadUrl } = this.element.data.video;
       return token && this.file && uploadUrl;
     }
   },
   methods: {
     upload() {
-      const { uploadUrl: url, token } = this.element.data;
+      const { uploadUrl: url, token } = this.element.data.video;
       return createUpload({ url, file: this.file, token })
         .then(({ id }) => {
           this.file = null;
           this.$emit('save', {
             ...this.element.data,
-            videoId: id,
-            status: ELEMENT_STATE.UPLOADED
+            video: {
+              ...this.element.data.video,
+              id,
+              status: ELEMENT_STATE.UPLOADED
+            }
           });
         })
         .catch(err => {
           this.$emit('save', {
             ...this.element.data,
-            error: get(err, 'response.data.error', DEFAULT_ERROR_MSG),
-            status: null,
-            fileName: null
+            video: {
+              ...this.element.data.video,
+              error: get(err, 'response.data.error', DEFAULT_ERROR_MSG),
+              status: null,
+              fileName: null
+            }
           });
         });
     }
   },
   watch: {
-    'element.data.uploadUrl'() {
+    'element.data.video.uploadUrl'() {
       if (this.isReadyToUpload) this.upload();
     }
   },
   mounted() {
-    this.$elementBus.on('save', ({ file }) => {
-      this.file = file;
-      this.$emit('save', {
-        ...this.element.data,
-        fileName: file.name,
-        error: null,
-        status: ELEMENT_STATE.UPLOADING
-      });
+    this.$elementBus.on('save', ({ video, caption }) => {
+      if (video) {
+        this.file = video.file;
+        delete video.file;
+      }
```

Comment:
Why do we need to do this?

Reply:
Because we don't want the `file` prop to be part of the asset.data object which is sent to the server on the `save` event.

Reply:
I manage to make it more readable using lodash methods.
```js
this.file = get(video, 'file', null);
const data = cloneDeep(this.element.data);
Object.assign(data.video, omit(video, ['file']));
Object.assign(data.caption, caption);
this.$emit('save', data);
```